airbyte-protocol/models/src/main/resources/airbyte_protocol/airbyte_protocol.yaml

```diff
@@ -70,14 +70,48 @@ definitions:
         type: string
   AirbyteStateMessage:
     type: object
-    additionalProperties: true
-    required:
-      - data
+    additionalProperties: false
     properties:
+      type:
+        "$ref": "#/definitions/AirbyteStateType"
       data:
-        description: "the state data"
-        type: object
-        existingJavaType: com.fasterxml.jackson.databind.JsonNode
+        description: "Deprecated"
+        "$ref": "#/definitions/AirbyteStateBlob"
+      global:
+        "$ref": "#/definitions/AirbyteStateBlob"
+      streams:
+        type: array
```

Comment:
why is this an array instead of a map?

Reply:
The map type doesn't exist in json schema. One version was using the additional properties for it but it was not clear: https://json-schema.org/draft/2020-12/json-schema-core.html#rfc.section.4.2.1

Reply:
would `"additionalProperties": { "type": AirbyteStateBlob }` work with `object` definition?

Reply:
I think it is worth exploring something like what Sherif proposed above. Otherwise we will need to need to sort through all of the streamStates in this array to find specific ones, rather than having direct access to a stream state given the stream name

Reply:
> would `"additionalProperties": { "type": AirbyteStateBlob }` work with `object` definition?

This would work but it is not explicit that the key is a string plus the field is gonna be named additionalProperties and not stream ([doc](https://github.com/joelittlejohn/jsonschema2pojo/wiki/Reference#additionalproperties)). We considered it at some point but abandoned it because of all the confusion it was bringing.
The tradeoff of rebuilding it as a map sounds better than having a field that it is not properly named and which can lead to have something unexpected in it.

Reply:
Maps in JSONSchema have been pretty sus. i.e. it seems like a lot of the codegen doesn't know what to do with them. I think the array is going to lead to less headache.

Lake makes a good point on the sorting, but I think we can handle it.